### PR TITLE
RFC: remote profile

### PR DESCRIPTION
**Pull request by dpwiz**

### CODE REVIEW

File: docs/rfcs/2023-09-12-remote-profile.md [hunk 1]
```diff
@@ -46,15 +46,20 @@ This can be solved in a few different ways:
...
 4. A desktop may create a temporary SMP queue and show its address. The mobile then submits its server data to it.
 
+Another option is to run the server on desktop and have mobile discover it with the help of QR code to get server identity and keys and then on the network via some protocol. Using a fixed address is suboptimal as most networks have dynamic IPs.
+
 ### Handshake
 
 The aim of this phase is to establish a TLS+cryptobox session.
 
+TLS could be complex as we need to generate self-signed certificates on desktop (if it acts like a server). A plaintext ws connection with cryptobox encryption could be sufficient initially? 
+
```
Comment (dpwiz) [inline docs/rfcs/2023-09-12-remote-profile.md:56]: Haskell-side, Ed25519 looks trivial to generate and pack into X509 cert. Post a fingerprint and you have it.

File: docs/rfcs/2023-09-12-remote-profile.md [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+### Disposal
+
+The session may have a lifetime that a desktop or a mobile may stipulate while preparing a session.
+Alternatively a mobile (or a desktop, why not) may signal that they're done here and no further activity should be going with the session parameters.
+
+## Proposed UX flow
+
+### On a mobile device
+
+1. A user opens sidebar and clicks "connect from desktop" in the "You" section, starting remote controller discovery.
```
Comment (epoberezkin) [inline docs/rfcs/2023-09-12-remote-profile.md:97]:
```suggestion
1. A user opens sidebar and clicks "use from desktop" in the "You" section, starting remote controller discovery.
```

File: docs/rfcs/2023-09-12-remote-profile.md [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+  * Discovery process starts UDP broadcast listener on application port (5226).
+  * A datagram containing remote controller fingerprint is checked against a list of pre-registered controller devices.
+    - If the device is already known, the host establishes connection and UI transitions into "connection status" window.
+    - For unknown/new broadcasts a fingerprint is displayed instead.
+    - If the datagram contains no valid fingerprint, it is ignored.
+2. Clicking on unknown device fingerprint in the list starts OOB handshake.
+  * UI enters "New device" window, displaying a fingerprint and asking to scan a QR code (or paste a link, like in the contact screen).
+  * A OOB data from the QR/link contains remote controller fingerprint and remote display name, which is stored in device DB.
+    - Fingerprint must match the announce.
+    - Display name can be changed.
```
Comment (epoberezkin) [inline docs/rfcs/2023-09-12-remote-profile.md:108]:
```suggestion
```

File: docs/rfcs/2023-09-12-remote-profile.md [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+  * A "connection status" window shows registered display name and current session status.
+  * Chat controller attempts to establish a remote session.
+    - A TCP connection is made to the address discovered.
+    - A TLS connection is instantiated, checking for remote CA fingerprint matches the previously established.
+    - A HTTP2 server is started on the mobile side of the TLS connection.
+    - The remote controller connects and subscribes for its output queue, marking the session established.
+  * For the duration of the remote session, the UI remains in the status window, preventing user interaction.
+    - This restriction may be lifted later.
+
+At any time a user may click on a "cancel" button and return to the main UI.
```
Comment (epoberezkin) [inline docs/rfcs/2023-09-12-remote-profile.md:120]: that should fully re-initialise UI state.

File: docs/rfcs/2023-09-12-remote-profile.md [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+In the "Network & servers" section of "Settings", there is an item to list all the registered remote controllers with buttons attached to *dispose* them one by one.
+*Disposing* a remote controller means its entry will be removed from database and any associated files deleted (there are none right now).
```
Comment (epoberezkin) [inline docs/rfcs/2023-09-12-remote-profile.md:123]: there isn't any associated files with remote controller, they are all app files.

File: docs/rfcs/2023-09-12-remote-profile.md [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+Future connection attempts from a disposed device would be treated exactly as from a previously-unknown device.
```
Comment (epoberezkin) [inline docs/rfcs/2023-09-12-remote-profile.md:124]: which means that desktop should be able to handle the scenario of connecting to known mobile that removed it - it should show qr code as a result of request from mobile.

File: docs/rfcs/2023-09-12-remote-profile.md [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+### On a desktop device
+
+1. A user opens sidebar and clicks "connect to mobile" in the "You" section.
+  * UI enters a "Select remote host" window asking user to pick an existing connection profile or generate a new one.
+2. When a new connection profile is requested by a user, UI transitions to "new remote host profile" window.
+  * A user may enter a device name that would be presented on a mobile device during OOB handshake, defaults to system device name.
```
Comment (epoberezkin) [inline docs/rfcs/2023-09-12-remote-profile.md:131]: it should be received from mobile.

File: docs/rfcs/2023-09-12-remote-profile.md [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+  * After submitting a private key is generated and a new X509 root CA certificate is produced and stored in device DB. Then the desktop proceeds to the connection screen.
+3. Clicking on an existing connection profile transitions UI to "connecting to remote host" window.
+  * A QR code / link is presented, containing the fingerprint of the CA CA stored for the selected profile.
+  * A new session certificate is derived from the CA.
+  * A TLS server is started using ephemeral session certificate.
+  * A UDP broadcast on port 5226 is started, sending the fingerprint.
+4. When incoming connection is established the UI transitions to "connected to remote host" window.
+  * Desktop chat controller establishes a remote session over the incoming connection (and prevents further connections).
+  * UI transitions to the "remote host" mode, shunting local profiles into background while keeping notifications coming.
+5. A user may open sidebar and click "disconnect from mobile" to close the session and return to local mode.
```
Comment (epoberezkin) [inline docs/rfcs/2023-09-12-remote-profile.md:141]: or connect to another.

File: docs/rfcs/2023-09-12-remote-profile.md [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+
+In the "Network & servers" section of "Settings", there is an item to list all the registered remote hosts with buttons attached to *dispose* them one by one.
```
Comment (epoberezkin) [inline docs/rfcs/2023-09-12-remote-profile.md:143]: this is not needed in this case, they will be under connected mobile devices.

File: docs/rfcs/2023-09-12-remote-profile.md (hunk 6 above)
Comment (dpwiz) [reply docs/rfcs/2023-09-12-remote-profile.md:124]: a desktop can always show a profile QR while waiting for connection.

**Pull request merged**